### PR TITLE
[WIP] Refactor libprefixed packages to multioutput

### PR DESCRIPTION
**Pull request by avnik**
###### Motivation for this change

Say "farewell" to long-lived ad-hoc hack, when we had two derivations for one package -- one for library-only version, second one with server/utils built.

See #14693 for problem history and details.

Done (Minimum to be done)
  - [x] ffado (complete, moved to own PR #42294 )
  - [x] ffado (aka ffadoFull ---  `ffado-mixer` moved to own derivation, and included in PR above)
  - [x] jack2 (waiting jack1 and own RP)
  - [  ] jack
  - [  ] pulseaudio (Work-in-progress, I still have some plans to split tools and daemon to own outputs)
  - [x] heimdal (done, own PR #42295 )

Not sure about this two, my expertise can be not enough. If I fail to refactor them, I'll create separate issues
  - [  ] krb5

At least volunters for testing would be needed, I haven't any ffado compatible hardware, and can test stuff only  for "just executed and shows version". Also I can't check darwin build myself.

I am sure, that is a mass-rebuild, so eventually this PR will be altered to be against staging. One of reasons, why I open PR early -- is prevend duplications of work, and use `ofborg` to look how much stuff affected.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by matthewbauer** [comment]
Hi! Thanks for doing this. I opened a similar PR in #39683 but you seem to have gotten further along. A few more that (maybe) can be removed:

- libv4l
- poppler_utils

File: pkgs/os-specific/linux/ffado/default.nix [hunk 1]
```diff
@@ -1,44 +1,57 @@
...
 
-  libOnly = prefix == "lib";
-
-  optLibjack2 = shouldUsePkg libjack2;
-  optDbus = shouldUsePkg dbus;
-  optDbus_cplusplus = shouldUsePkg dbus_cplusplus;
-  optAlsaLib = shouldUsePkg alsaLib;
+# Mixer/tools dependencies
+# FIXME: can we build mixer separately? Then we can farewell `libOnly` mode
+# (otherwise we have perfect loop via qt -> pulseaudio -> jack -> ffado -> qt)
```
Comment (matthewbauer) [inline pkgs/os-specific/linux/ffado/default.nix]: We can probably override ffado in jack's callPackage-jack shouldn't need Qt.

File: pkgs/misc/jackaudio/default.nix [hunk 2]
```diff
@@ -17,12 +16,10 @@ let
   inherit (python2Packages) python dbus-python;
   shouldUsePkg = pkg: if pkg != null && pkg.meta.available then pkg else null;
 
-  libOnly = prefix == "lib";
-
   optDbus = if stdenv.isDarwin then null else shouldUsePkg dbus;
-  optPythonDBus = if libOnly then null else shouldUsePkg dbus-python;
-  optLibffado = if libOnly then null else shouldUsePkg libffado;
-  optAlsaLib = if libOnly then null else shouldUsePkg alsaLib;
+  optPythonDBus = shouldUsePkg dbus-python;
```
Comment (matthewbauer) [inline pkgs/misc/jackaudio/default.nix]: You can probably get rid of shouldUsePkg. I think they are officially deprecated (just use dbus-python directly).

**Comment by oxij** [comment]
Let me note here (too) that I think that some parts of `libOnly` are useful https://github.com/NixOS/nixpkgs/pull/39683#issuecomment-385682730 and that the problem, as I see it, is not only that `libOnly` can be implemented with multiple outputs, but that it also is not a use-flag, i.e. derivations that use it are not monoidal https://github.com/NixOS/nixpkgs/pull/39683#issuecomment-385690025 and monoidal derivations are FTW.

File: pkgs/os-specific/linux/ffado/default.nix (hunk 1 above)
Comment (avnik) [reply pkgs/os-specific/linux/ffado/default.nix]: ffado' mixer is actually separate app, which have completely own dependencies (dbus-python and PyQt), so I feel  separate expression may be better solution. But would be nice to hear someone who use ffado.

File: pkgs/misc/jackaudio/default.nix (hunk 2 above)
Comment (avnik) [reply pkgs/misc/jackaudio/default.nix]: It don't used outside of mixer, so if we move mixer to own app, we can drop dbus-python in main ffado.

**Comment by avnik** [comment]
@oxij  libOnly here is definelly not use-flag.

Other question to all -- I had intention to split `daemon` part of pulseaudio from client binaries (to `daemon` and `bin` outputs respectively) -- does it a good idea, or too radical?  Then we can drop some useflags from pulseaudio -- extra dependencies (like jack/ffado) needed only for daemon, but not for client utilities like pactl/pacmd/....

**Comment by avnik** [comment]
Some work, which already done kicked to own PRs, just updated this issue to keep progress tracking (still use it as WIP branch)

**Comment by mmahut** [comment]
What is the status of this pull request?

**Comment by avnik** [comment]
@mmahut  `ffado` and `heimdal`  done and merged. Pulseaudio still pending,  I'll try to rebase my patches and push them to own PR,  also I feel worth to migrate pulseaudio package to using meson, instead autoconf.

**Comment by avnik** [comment]
Actually only  pulseaudio refactoring pending, so I can close it

**Pull request closed without merging**